native/java/org/jpype/JPypeContext.java

```diff
@@ -100,17 +104,131 @@ public static JPypeContext createContext(long context, ClassLoader bootLoader)
     instance.typeManager.init();
     instance.referenceQueue.start();
 
+    // Install a shutdown hook to clean up Python resources.
+    Runtime.getRuntime().addShutdownHook(new Thread(new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        instance.watchShutdown();
+      }
+    }));
+
     return instance;
   }
 
+  /**
+   * Shutdown and remove all Python resources.
+   *
+   * This hook is only called after the last user thread has died. Thus the only
+   * remaining connections are proxies that were attached to the JVM shutdown
+   * hook, the reference queue, and the typemanager.
+   *
+   * This routine will try to take out the last connections in an orderly
+   * fashion. Inherently this is a very dangerous time as portions of Java have
+   * already been deactivated.
+   */
+  private void watchShutdown()
+  {
+    // Try to yield in case there is a race condition.  The user
+    // may have installed a shutdown hook, but we cannot verify
+    // the order that shutdown hook threads are executed.  Thus we will
+    // try to intentionally lose the race.
+    //
+    // This will only occur if something registered a shutdown hook through
+    // a Java API.  Those registered though the JPype API will be joined
+    // manually.
+    for (int i = 0; i < 5; i++)
+    {
+      try
+      {
+        Thread.sleep(1);
+        Thread.yield();
+      } catch (InterruptedException ex)
+      {
+      }
+    }
+
+    // Execute any used defined shutdown hooks registered with JPype.
+    if (!this.shutdownHooks.isEmpty())
+    {
+      for (Thread thread : this.shutdownHooks)
+      {
+        thread.start();
+      }
+      for (Thread thread : this.shutdownHooks)
+      {
+        try
+        {
+          thread.join();
+        } catch (InterruptedException ex)
+        {
+        }
+      }
+    }
+
+    // Disable all future calls to proxies
+    this.shutdownFlag.incrementAndGet();
+
+    // Wait for any unregistered proxies to finish so that we don't yank
+    // the rug out from under them result in a segfault.
```

Comment:
there seems to be a lot of things which can go horribly wrong :)

Reply:
Yeah.  There are hoards of things that can break.   Which is way I had to try to redirect them to a list I know so that I can join all of them.